### ISSUE TITLE
[Bug]: The TPCH 10G  q21 was hung on EKS with 2*cn (pull mode)

### ISSUE THREAD

**Issue by aressu1985** [matrixorigin/matrixone#8996]
### Is there an existing issue for the same bug?

- [X] I have checked the existing issues.

### Environment

```Markdown
- Version or commit-id (e.g. v0.1.0 or 8b23a93): b7b1f1f
- Hardware parameters:16C 64G  2*cn 1*dn 3*logservice
- OS type:
- Others:
```

### Actual Behavior

In daily regression test  on eks , the Q21 of tpch 10G scale was hung , mo can not return correct result for about several hours.
<img width="826" alt="image" src="https://user-images.githubusercontent.com/47846308/231925216-0e458ea7-3190-4e37-b2cb-a2db569aa29b.png">

[tpch_q21_hung_cn0.txt](https://github.com/matrixorigin/matrixone/files/11228151/tpch_q21_hung_cn0.txt)
[tpch_q21_hung_cn1.txt](https://github.com/matrixorigin/matrixone/files/11228158/tpch_q21_hung_cn1.txt)

### Steps to Reproduce

```Markdown
1. run a mo cluster on eks with (2*cn, 1*dn, 3*log)
2. load tpch 10g data 
3. run tpch 10g q21 query

or 

1. git clone https://github.com/matrixorigin/mo-auto-test.git
2. modify .github/workflows/regression-on-distributed-mode.yml, 
  set JobName.default to tpch_benchmark_test
  set RunTimes to 1
3. make a pr to merge, and the job will automatically run the tpch 10g queries on eks
```

**Comment by aressu1985** [comment]
mysql> explain select
    ->     s_name,
    ->     count(*) as numwait
    -> from
    ->     supplier,
    ->     lineitem l1,
    ->     orders,
    ->     nation
    -> where
    ->     s_suppkey = l1.l_suppkey
    ->     and o_orderkey = l1.l_orderkey
    ->     and o_orderstatus = 'F'
    ->     and l1.l_receiptdate > l1.l_commitdate
    ->     and exists (
    ->         select
    ->             *
    ->         from
    ->             lineitem l2
    ->         where
    ->             l2.l_orderkey = l1.l_orderkey
    ->             and l2.l_suppkey <> l1.l_suppkey
    ->     )
    ->     and not exists (
    ->         select
    ->             *
    ->         from
    ->             lineitem l3
    ->         where
    ->             l3.l_orderkey = l1.l_orderkey
    ->             and l3.l_suppkey <> l1.l_suppkey
    ->             and l3.l_receiptdate > l3.l_commitdate
    ->     )
    ->     and s_nationkey = n_nationkey
    ->     and n_name = 'BRAZIL'
    -> group by
    ->     s_name
    -> order by
    ->     numwait desc,
    ->     s_name
    -> limit 100
    -> ;
+------------------------------------------------------------------------------------------------------+
| QUERY PLAN                                                                                           |
+------------------------------------------------------------------------------------------------------+
| Project                                                                                              |
|   ->  Sort                                                                                           |
|         Sort Key: count(1) DESC, supplier.s_name INTERNAL                                            |
|         Limit: 100                                                                                   |
|         ->  Aggregate                                                                                |
|               Group Key: supplier.s_name                                                             |
|               Aggregate Functions: starcount(1)                                                      |
|               ->  Join                                                                               |
|                     Join Type: SEMI                                                                  |
|                     Join Cond: (l2.l_orderkey = l1.l_orderkey), (l2.l_suppkey <> l1.l_suppkey)       |
|                     ->  Table Scan on tpch_10g.lineitem                                              |
|                     ->  Join                                                                         |
|                           Join Type: ANTI                                                            |
|                           Join Cond: (l3.l_orderkey = l1.l_orderkey), (l3.l_suppkey <> l1.l_suppkey) |
|                           ->  Table Scan on tpch_10g.lineitem                                        |
|                                 Filter Cond: (l3.l_receiptdate > l3.l_commitdate)                    |
|                           ->  Join                                                                   |
|                                 Join Type: INNER                                                     |
|                                 Join Cond: (orders.o_orderkey = l1.l_orderkey)                       |
|                                 ->  Table Scan on tpch_10g.orders                                    |
|                                       Filter Cond: (orders.o_orderstatus = 'F')                      |
|                                 ->  Join                                                             |
|                                       Join Type: INNER                                               |
|                                       Join Cond: (supplier.s_suppkey = l1.l_suppkey)                 |
|                                       ->  Table Scan on tpch_10g.lineitem                            |
|                                             Filter Cond: (l1.l_receiptdate > l1.l_commitdate)        |
|                                       ->  Join                                                       |
|                                             Join Type: INNER                                         |
|                                             Join Cond: (supplier.s_nationkey = nation.n_nationkey)   |
|                                             ->  Table Scan on tpch_10g.supplier                      |
|                                             ->  Table Scan on tpch_10g.nation                        |
|                                                   Filter Cond: (nation.n_name = 'BRAZIL')            |
+------------------------------------------------------------------------------------------------------+
32 rows in set (0.18 sec)

**Comment by aressu1985** [comment]
and q22, q4 also are hung and mo can not return resluts in 5 minutes;

q4:
<img width="663" alt="image" src="https://user-images.githubusercontent.com/47846308/231937127-46d6e4c5-e2fe-498d-bf31-cf6d841784ce.png">

q22:
<img width="1408" alt="image" src="https://user-images.githubusercontent.com/47846308/231937275-f4608aca-3476-47b9-bae6-154193f588ff.png">

**Comment by m-schen** [comment]
can modify `isSameCN` method to return parts1[0] == parts2[0] && parts1[1] == parts2[1] to Simulate a multi cn environment locally

**Comment by dongdongyang33** [comment]
here are 7 need-notify-pipeline but only 3 of them send the notify message
<img width="1540" alt="image" src="https://user-images.githubusercontent.com/47596332/232015816-006497a8-81b4-47ec-b543-b2c9f09cb23c.png">

**Comment by aunjgr** [comment]
can be reproduced on multi CN. Still investigating the reason.

**Comment by dongdongyang33** [comment]
cause by missing `RightAnti` type when decode scope which received from remote

**Comment by aunjgr** [comment]
> cause by missing `RightAnti` type when decode scope which received from remote.
> 
> <img alt="image" width="852" src="https://user-images.githubusercontent.com/47596332/232205703-b09ca543-d0e1-47af-bf07-a9458f4d2dbd.png">
> 
> But the error message seems lost, so the query hung.

Thanks for this hint. `RightSemi` is also missing in `convertToVmInstruction`.

Now we have too much `Join` operators, 15 in total. And adding one means changing many type switches distributed in multiple files under multiple dir. That's too error-prone. We should merge them into only two: `HashJoin` and `NestedLoopJoin`, maybe after the massive operator/expression refactor.

**Comment by aressu1985** [comment]
this issue is fixed , but cause another issue #9013 

this issue can be fixed

**Comment by aressu1985** [comment]
testing
